### PR TITLE
Add GEMM to benchmark suite, and refactor

### PR DESCRIPTION
**Pull request by AdamHarries**

### CODE REVIEW

File: bench/syclblas/syclblas_benchmark_level2.cpp [hunk 1]
```diff
@@ -25,17 +25,75 @@
...
+
+  benchmark<>::flops_units_t flops = benchmark<>::measure(
+      reps, n_fl_ops,
+      [&]() {
+        auto event = _gemv(ex, *t_str, m, n, alpha, m_a_gpu, m, v_b_gpu, incX,
+                           beta, v_c_gpu, incY);
+        ex.wait(event);
+      });
+
+      auto event = ex.copy_to_host(v_c_gpu, c_v_gpu_result.data(), rlen);
```
Comment (Ruyk) [inline bench/syclblas/syclblas_benchmark_level2.cpp]: why the indent here?

File: bench/syclblas/syclblas_benchmark_level2.cpp [hunk 2]
```diff
@@ -25,17 +25,75 @@
...
+  ex.template deallocate<ScalarT>(v_c_gpu);
+
+  return flops;
+}
+
+SUITE(ADD(gemv))
+
+
+// TODO: This doesn't seem to work with transposed matrices? 
+// Is it a problem with the routine, or the inputs?
```
Comment (Ruyk) [inline bench/syclblas/syclblas_benchmark_level2.cpp]: Create a github issue for this

File: bench/blas_benchmark.hpp [hunk 3]
```diff
@@ -47,8 +51,37 @@ ScalarT *new_data(size_t size, bool initialized = true) {
   }
   return v;
 }
+
+template <typename ScalarT>
+ScalarT *new_const_data(size_t size, ScalarT value = 0) {
+  ScalarT *v = new ScalarT[size];
```
Comment (Ruyk) [inline bench/blas_benchmark.hpp]: do we need a raw pointer? can we use unique_ptr at least?

File: bench/blas_benchmark.hpp [hunk 4]
```diff
@@ -100,28 +133,58 @@ struct benchmark {
   }
 };
 
+// why on earth do we need this??
```
Comment (Ruyk) [inline bench/blas_benchmark.hpp]: good question

File: bench/range.hpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
+#ifndef RANGE_HPP
```
Comment (Ruyk) [inline bench/range.hpp]: file is missing license header

File: bench/range.hpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+      : v(_low), low(_low), high(_high), mult(_mult) {}
+  SizeRange(const SizeRange& sr)
+      : v(sr.v), low(sr.low), high(sr.high), mult(sr.mult) {}
+  T peek() { return v; }
+  T yield() {
+    T r = v;    // cache the current value
+    v *= mult;  // increment
+    return r;   // return
+  }
+  bool finished() {
```
Comment (Ruyk) [inline bench/range.hpp]: missing final keyword

File: bench/range.hpp [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+template <typename Range1, typename Range2, typename Range3, typename Range4,
+          typename Range5>
+Range5D<Range1, Range2, Range3, Range4, Range5> nd_range(Range1 r1, Range2 r2,
+                                                         Range3 r3, Range4 r4,
+                                                         Range5 r5) {
+  return Range5D<Range1, Range2, Range3, Range4, Range5>(r1, r2, r3, r4, r5);
+}
+
+#endif  // include guard
```
Comment (Ruyk) [inline bench/range.hpp]: is this coming from elsewhere? is there a link to the source?

File: bench/syclblas/syclblas_benchmark_level3.cpp [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        << std::get<4>(params);
+  return fname.str();
+}
+
+BENCHMARK(gemm, blas_level_3) {
+  using ScalarT = ElemT;
+
+  size_t m = std::get<0>(params);
+  size_t k = std::get<1>(params);
+  size_t n = std::get<2>(params);
```
Comment (Ruyk) [inline bench/syclblas/syclblas_benchmark_level3.cpp]: the three above can be const

File: bench/syclblas/syclblas_benchmark_level3.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  char const *t_a = std::get<3>(params);
+  char const *t_b = std::get<4>(params);
```
Comment (Ruyk) [inline bench/syclblas/syclblas_benchmark_level3.cpp]: why not std::string?

File: bench/syclblas/syclblas_benchmark_level2.cpp (hunk 2 above)
Comment (AdamHarries) [reply bench/syclblas/syclblas_benchmark_level2.cpp]: After re-running the code, it does actually seem to work for transposed inputs. I think I managed to fix the bug, but didn't remove the comment.

File: bench/blas_benchmark.hpp (hunk 3 above)
Comment (AdamHarries) [reply bench/blas_benchmark.hpp]: It turns out that we don't actually need these functions at all - I think they're left over from an earlier refactoring pass. Removing.

File: bench/range.hpp (hunk 7 above)
Comment (AdamHarries) [reply bench/range.hpp]: Do you mean the source code in range.hpp? No, it's all written by me...

File: bench/syclblas/syclblas_benchmark_level3.cpp (hunk 9 above)
Comment (AdamHarries) [reply bench/syclblas/syclblas_benchmark_level3.cpp]: the BLAS interface expects an old c-style string, so avoiding converting to/from string saves a little cognitive overhead.

Comment (Ruyk) [reply bench/syclblas/syclblas_benchmark_level3.cpp]: Ok, let it be...

**Pull request merged**